protocol/arq.go

```diff
@@ -33,83 +33,79 @@ func (arq *goBackNArq) addExtension(extension Connector) {
 	arq.extension = extension
 }
 
-func nextSegment(currentIndex int, sequenceNum uint32, buffer []byte) (int, segment) {
-	var next = currentIndex + dataChunkSize
-	var flag byte = 0
-	if currentIndex == 0 {
-		flag |= flagSyn
-	}
-	if next >= len(buffer) {
-		flag |= flagEnd
-		next = len(buffer)
-	}
-	return next, createFlaggedSegment(sequenceNum, flag, buffer[currentIndex:next])
-}
-
-func (arq *goBackNArq) queueSegments(buffer []byte) {
+func (arq *goBackNArq) setupSegmentsQueues(buffer []byte) {
 	var currentIndex = 0
 	var segmentCount = 0
 	var seg segment
 	var sequenceNumber = initialSequenceNumber()
 	arq.initialSequenceNumber = sequenceNumber
-	arq.lastSegmentAcked = -1
+	arq.lastAckedSequenceNumber = -1
 	for {
-		currentIndex, seg = nextSegment(currentIndex, sequenceNumber, buffer)
-		arq.segmentQueue.Enqueue(seg)
+		currentIndex, seg = peekFlaggedSegmentOfBuffer(currentIndex, sequenceNumber, buffer)
+		arq.readyToSendSegmentQueue.Enqueue(seg)
 		segmentCount++
 		sequenceNumber++
-		if seg.isFlaggedAs(flagEnd) {
+		if seg.isFlaggedAs(FlagEND) {
 			break
 		}
 	}
-	arq.segmentWriteBuffer = make([]*segment, segmentCount)
+	arq.notAckedSegmentQueue = make([]*segment, segmentCount)
 }
 
-func (arq *goBackNArq) acksPending() bool {
-	return arq.lastSegmentAcked != -1 && arq.lastSegmentAcked < len(arq.segmentWriteBuffer)-1
+func (arq *goBackNArq) hasAcksPending() bool {
+	return arq.lastAckedSequenceNumber != -1 && arq.lastAckedSequenceNumber < len(arq.notAckedSegmentQueue)-1
 }
 
 func (arq *goBackNArq) Write(buffer []byte) (statusCode, int, error) {
+
+	if buffer == nil {
+		panic("Paramater is nil")
+	}
+
 	arq.writeMutex.Lock()
-	defer arq.writeMutex.Unlock()
 
 	if len(buffer) > 0 {
-		if !arq.segmentQueue.IsEmpty() || arq.acksPending() {
+		if !arq.readyToSendSegmentQueue.IsEmpty() || arq.hasAcksPending() {
 			return pendingSegments, 0, nil
 		}
-		arq.queueSegments(buffer)
+		arq.setupSegmentsQueues(buffer)
 	}
-	arq.requeueTimedOutSegments()
-	status, n, err := arq.writeQueuedSegments()
-	return status, n, err
+
+	arq.writeMutex.Unlock()
```

Comment:
Das wird zu Fehler führen. Der Hauptgrund für den defer oben war, das es retunrs in der methode gibt. Einen mutex unlock würde ich immer mit defer aufrufen aus diesem Grund.